asset/asset.go

```diff
@@ -247,6 +247,71 @@ func DecodeGenesis(r io.Reader) (Genesis, error) {
 	return gen, err
 }
 
+// Specifier is a type that can be used to specify an asset by its ID, its asset
+// group public key, or both.
+type Specifier struct {
+	// id is the asset ID.
+	id fn.Option[ID]
+
+	// groupKey is the asset group public key.
+	groupKey fn.Option[btcec.PublicKey]
+}
+
+// NewSpecifier creates a new specifier that specifies an asset by its ID and
+// its group public key.
+func NewSpecifier(id ID, groupPubKey btcec.PublicKey) Specifier {
```

Comment:
To avoid needing to do `if a.groupKey != nil {` in a couple of places, should we have two constructors instead, that take the group key as a pointer each?

What I mean:
```
// NewSpecifier creates a new specifier that specifies an asset by its ID and
// its group public key.
func NewSpecifier(id ID, groupPubKey *btcec.PublicKey) Specifier {
	s := Specifier{
		id: fn.Some(id),
	}

	if groupPubKey != nil {
		s.groupKey = fn.Some(*groupPubKey)
	}

	return s
}

// NewSpecifierWithGroup creates a new specifier that specifies an asset by its
// ID and its group key.
func NewSpecifierWithGroup(id ID, groupKey *GroupKey) Specifier {
	s := Specifier{
		id: fn.Some(id),
	}
	
	if groupKey != nil {
		s.groupKey = fn.Some(groupKey.GroupPubKey)
	}
	
	return s
}
```

Those two should simplify the code in a lot of places.

Reply:
I'm not a fan of this sort of signature:
```
func NewSpecifier(id ID, groupPubKey *btcec.PublicKey) Specifier
```
It doesn't make it clear that `groupPubKey` is optional. Pointer doesn't always mean optional. So as a developer using `NewSpecifier` I need to read the function body or the doc to figure out if I have to set that arg. Often times the function doc does not help.

That's why I went with:
```
func NewSpecifier(id ID, groupPubKey btcec.PublicKey) Specifier
func NewSpecifierFromId(id ID) Specifier
func NewSpecifierFromGroupKey(groupPubKey btcec.PublicKey) Specifier
```

---

I find `NewSpecifier` and  `NewSpecifierWithGroup` in your examples confusing because the names could be swapped without any effect. `NewSpecifierWithGroup` is just getting to `groupKey.GroupPubKey`, so it seems redundant to me, that field could be passed in at the call site?

---

How do you construct an asset specifier from a group key alone with your suggested `NewSpecifier` and  `NewSpecifierWithGroup`? I think that might be useful when we're doing RFQ for any asset in a group, for example. Or creating a tap address for any asset in a group. So I think the specifier should be able to refer to a group of assets without needing to refer to a single asset.

Reply:
Okay, I see how the naming might be confusing... But you see my point of trying to avoid all the `if groupKey != nil {` conditions all over the code base?
So maybe we can make it more obvious by putting the "optional" into the name? So `NewSpecifierGroupPubKeyOptional(id ID, groupPubKey *btcec.PublicKey)` and `NewSpecifierGroupKeyOptional(id ID, groupKey *asset.GroupKey)`?

Reply:
I've added those constructors in the latest commits.